[css] Add a "did you mean" suggestion when an unknown CSS property is spotted

DrBenton opened the pull request:
So using `bckgroundu: red` in a CSS file reports _"unknown declaration 'bckgroundu'; did you mean 'background'?"_, while using a very wrongly typed CSS property doesn't suggest any.
![Screenshot from 2022-04-29 12-14-07](https://user-images.githubusercontent.com/722388/165936281-82b027a6-b1d3-4cbc-ae12-00a4c7ef987e.png)

(yeah, I'm afraid italics markups rendering is not as good in the Gnome terminal than in iTerm :sweat_smile: )

closes #431

DrBenton commented on src/textual/css/_help_text.py:
```diff
@@ -127,7 +128,31 @@ def _spacing_examples(property_name: str) -> ContextSpecificBullets:
 
 
-def spacing_wrong_number_of_values(
+def property_invalid_value_help_text(
```
a new HelpText factory, added to the ones already created by Darren :-)

DrBenton commented on src/textual/css/_help_text.py:
```diff
@@ -158,7 +184,7 @@ def spacing_wrong_number_of_values(
 
 
-def spacing_invalid_value(
+def spacing_invalid_value_help_text(
```
While I was exploring this "HelpText factories" machinery I noticed that contrary to the others the first two ones didn't have the `help_text` suffix - for teh sake of consistency, I added it :slightly_smiling_face:

DrBenton commented on src/textual/css/_styles_builder.py:
```diff
@@ -332,14 +349,20 @@ def _process_space(self, name: str, tokens: list[Token]) -> None:
+                    self.error(
+                        name,
+                        token,
+                        spacing_invalid_value_help_text(name, context="css"),
```
just a "Black formatting" consequence of adding the missing `_help_text` suffix to these function names

DrBenton commented on tests/css/test_help_text.py:
```diff
@@ -1,10 +1,22 @@
-    string_enum_help_text, color_property_help_text, border_property_help_text, layout_property_help_text, \
-    docks_property_help_text, dock_property_help_text, fractional_property_help_text, offset_property_help_text, \
-    align_help_text, offset_single_axis_help_text, style_flags_property_help_text
+from textual.css._help_text import (
```
here again, just the consequences of adding the `_help_text` suffix on these 2 function names

willmcgugan commented on src/textual/css/_help_text.py:
```diff
@@ -128,7 +128,32 @@ def _spacing_examples(property_name: str) -> ContextSpecificBullets:
+        suggested_property_name = _contextualize_property_name(
+            suggested_property_name, context
+        )
+        bullets.append(Bullet(f"Did you mean [i]{suggested_property_name}[/]?"))
```
I think we might want to drop the italics and use quotes instead. i.e. `Did you mean "background"?`

DrBenton replied:
Done!
![Screenshot from 2022-05-03 15-22-07](https://user-images.githubusercontent.com/722388/166471761-eced1754-2c68-4008-b971-d8232cd80f96.png)

(I've left the first one with italics though, as it's the style used by our other HelpTexts when they display a CSS property name)

willmcgugan commented on src/textual/css/_styles_builder.py:
```diff
@@ -717,3 +744,9 @@ def process_align_vertical(self, name: str, tokens: list[Token]) -> None:
     process_content_align_horizontal = process_align_horizontal
     process_content_align_vertical = process_align_vertical
+
+    def _suggested_property_name_for_rule(self, rule_name: str) -> str | None:
```
Docstring por fovor.

willmcgugan commented on src/textual/css/_styles_builder.py:
```diff
@@ -717,3 +744,9 @@ def process_align_vertical(self, name: str, tokens: list[Token]) -> None:
     process_content_align_vertical = process_align_vertical
+
+    def _suggested_property_name_for_rule(self, rule_name: str) -> str | None:
+        possible_matches = get_close_matches(
```
Can we factor this function in to a new module. Probably `suggestions.py`. 

I'm thinking two functions, one for a single suggestion, and another for multiple. With signatures along these lines:

```python
def get_suggestion(word:str, possible_words:list[str]) -> str | None:
```

```python
def get_suggestions(word:str possible_words:list[str], count:int) -> list[str]:
```

willmcgugan commented on src/textual/css/_styles_builder.py:
```diff
@@ -84,24 +87,38 @@ def add_declaration(self, declaration: Declaration) -> None:
+            self.error(declaration.name, declaration.token, str(error))
+
+    @lru_cache(maxsize=None)
+    def _processable_rule_names(self) -> frozenset[str]:
```
Best to start with a verb, suggest `_get_processable_rule_names`

Docstring please.

DrBenton replied:
Ah, I was not sure about that one, as the two naming conventions exist in the wild! :smile: 
In Symfony for example we use `$request->getClientIp()` to get the IP address from a HTTP request, whereas it's `$request->ip()` in Laravel.
I'll change to the "getter" style! :+1: :slightly_smiling_face:

willmcgugan commented on src/textual/css/_styles_builder.py:
```diff
@@ -63,27 +68,37 @@ def add_declaration(self, declaration: Declaration) -> None:
+
+    @lru_cache(maxsize=None)
+    def _processable_rule_names(self) -> frozenset[str]:
+        return frozenset(
```
I suspect it doesn't need to be a set. Correct me if I'm wrong but we will only ever iterate through the container, and never check for set membership. In that case, a plain old list is probably better.

DrBenton replied:
It doesn't need to be a set indeed - I chose this data structure only because it seems to be the one with the fastest lookup. But as this is something that is only triggered when fatal errors happen I guess performance is not really a concern there.
I'll change that to a list! :slightly_smiling_face:  :+1:

willmcgugan reviewed (changes requested):
A few requests

DrBenton commented:
@willmcgugan Thanks for your review! Hopefully addressed your comments in [26f138e](https://github.com/Textualize/textual/pull/445/commits/26f138e69be49f33fe7ff72cebbb51d617a6338f) :crossed_fingers:  :slightly_smiling_face:

DrBenton commented on src/textual/suggestions.py:
```diff
@@ -0,0 +1,24 @@
+
+
+def get_suggestion(word: str, possible_words: Sequence[str]) -> str | None:
+    """
```
Ah, true, I forgot that we were using this format sorry! I'll do that! :+1:

DrBenton replied:
Done! [a26822b](https://github.com/Textualize/textual/pull/445/commits/a26822b257e187969f9f2dcef0e17af88650995d)

I'll try to better document the stuff I'll add from now on :slightly_smiling_face:

willmcgugan commented:
très bien!
